### ISSUE TITLE
Download-Artifacts throws Mutex exception "Access to the path ... is denied"

### ISSUE THREAD

**Issue by b-zijlstra** [microsoft/navcontainerhelper#3164]
**Scenario**
I have a VM with 6 Azure DevOps Agents that share an Agent Pool. In that pool of 6 agents I am running multiple pipelines simultaneously that need to download the same BC artifacts. If the pipelines start running at the same time, the Download-Artifacts step fails most of the time. This is also the case when I use a matrix strategy in a single pipeline with maxParallel > 1 (i.e. it is running the same steps from multiple agents on the server at the same time).  I get the error when the mutex is created. All agents are running under the same service account that has admin privileges and full control over the bcartifacts.cache folder.
```
New-Object : Exception calling ".ctor" with "2" argument(s): "Access to the path 'dl-bcartifacts.azureedge.net_onprem_21.1.48363.48638_w1' is denied."
```

**This Pipeline will fail most of the time, unless I increase the Start-Sleep delay to e.g. 1000 milliseconds**
```
trigger: none
stages:
  - stage: Download
    jobs:
      - job: DownloadConcurrent
        strategy:
          maxParallel: "2"
          matrix:
            RunOne:
              delay: 0
            RunTwo:
              delay: 250
        steps:
          - task: PowerShell@2
            inputs:
              pwsh: true
              targetType: inline
              script: |
                Start-Sleep -Milliseconds $(delay)
                $artifactUrl = Get-BCArtifactUrl -type "OnPrem" -version "21.1" -country "w1"
                Download-Artifacts -artifactUrl $artifactUrl

```

**Output of the PowerShell step for the failing Run**
```
Starting: PowerShell
==============================================================================
Task         : PowerShell
Description  : Run a PowerShell script on Linux, macOS, or Windows
Version      : 2.226.1
Author       : Microsoft Corporation
Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/powershell
==============================================================================
Generating script.
========================== Starting Command Output ===========================
"C:\Program Files\PowerShell\7\pwsh.exe" -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'C:\agent_build01\_work\_temp\b04648b8-76d6-4ef3-af31-7523a49777d7.ps1'"
BcContainerHelper version 5.0.4
BC.HelperFunctions emits usage statistics telemetry to Microsoft
Running on PowerShell 7
Download-Artifacts Telemetry Correlation Id: 219ea7e0-d9c4-473c-9a7e-b813fed0b982
New-Object : Exception calling ".ctor" with "2" argument(s): "Access to the path 'dl-bcartifacts.azureedge.net_onprem_21.1.48363.48638_w1' is denied."
At C:\Program Files\PowerShell\Modules\BcContainerHelper\5.0.4\Artifacts\Download-Artifacts.ps1:48 char:17
+ …  $appMutex = New-Object System.Threading.Mutex($false, $appMutexName)
+                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : InvalidOperation: (:) [New-Object], MethodInvocationException
+ FullyQualifiedErrorId : ConstructorInvokedThrowException,Microsoft.PowerShell.Commands.NewObjectCommand
##[error]PowerShell exited with code '1'.
Finishing: PowerShell
```
**Output of the PowerShell step for the successful Run**
```
Starting: PowerShell
==============================================================================
Task         : PowerShell
Description  : Run a PowerShell script on Linux, macOS, or Windows
Version      : 2.226.1
Author       : Microsoft Corporation
Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/powershell
==============================================================================
Generating script.
========================== Starting Command Output ===========================
"C:\Program Files\PowerShell\7\pwsh.exe" -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'C:\agent_build02\_work\_temp\dbcd514d-87cf-49ba-be1c-70dc5d39373d.ps1'"
BcContainerHelper version 5.0.4
BC.HelperFunctions emits usage statistics telemetry to Microsoft
Running on PowerShell 7
c:\bcartifacts.cache\onprem\21.1.48363.48638\w1
Finishing: PowerShell
```

**Interestingly, the following pipeline does not seem to fail**
```
trigger: none
stages:
  - stage: Download
    jobs:
      - job: Download
        steps:
          - task: PowerShell@2
            inputs:
              pwsh: true
              targetType: inline
              script: |
                $artifactUrl = Get-BCArtifactUrl -type "OnPrem" -version "21.1" -country "w1"
                @($artifactUrl , $artifactUrl, $artifactUrl , $artifactUrl , $artifactUrl , $artifactUrl) | ForEach-Object -Parallel { Download-Artifacts -artifactUrl $_; } -ThrottleLimit 6
```
```
Starting: PowerShell
==============================================================================
Task         : PowerShell
Description  : Run a PowerShell script on Linux, macOS, or Windows
Version      : 2.226.1
Author       : Microsoft Corporation
Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/powershell
==============================================================================
Generating script.
========================== Starting Command Output ===========================
"C:\Program Files\PowerShell\7\pwsh.exe" -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'C:\agent_build01\_work\_temp\3e159c7e-0f0a-4489-965e-b87dc8e3cbd3.ps1'"
BcContainerHelper version 5.0.4
BC.HelperFunctions emits usage statistics telemetry to Microsoft
Running on PowerShell 7
BcContainerHelper version 5.0.4
BcContainerHelper version 5.0.4
BcContainerHelper version 5.0.4
BcContainerHelper version 5.0.4
BcContainerHelper version 5.0.4
BC.HelperFunctions emits usage statistics telemetry to Microsoft
BcContainerHelper version 5.0.4
BC.HelperFunctions emits usage statistics telemetry to Microsoft
BC.HelperFunctions emits usage statistics telemetry to Microsoft
BC.HelperFunctions emits usage statistics telemetry to Microsoft
BC.HelperFunctions emits usage statistics telemetry to Microsoft
BC.HelperFunctions emits usage statistics telemetry to Microsoft
Running on PowerShell 7
Running on PowerShell 7
Running on PowerShell 7
Running on PowerShell 7
Running on PowerShell 7
Running on PowerShell 7
c:\bcartifacts.cache\onprem\21.1.48363.48638\w1
c:\bcartifacts.cache\onprem\21.1.48363.48638\w1
c:\bcartifacts.cache\onprem\21.1.48363.48638\w1
c:\bcartifacts.cache\onprem\21.1.48363.48638\w1
c:\bcartifacts.cache\onprem\21.1.48363.48638\w1
c:\bcartifacts.cache\onprem\21.1.48363.48638\w1
Finishing: PowerShell
```
It seems to me that the error only occurs when code is running concurrently on two different agents. Any idea what could cause this? I can partially work around this by not running matrix strategies, however builds will still fail if two pipelines happen to start running at the same time.

Note: Usually I'm not running Download-Artifacts directly. In my actual pipelines I'm getting the error when calling New-BcContainer or when using e.g. the ALOpsAppCompiler@2 build step.

**Comment by freddydk** [comment]
Can you repro this on your own machine?
or on the Azure VM hosting your agents - by simply using PowerShell?
I cannot:
![image](https://github.com/microsoft/navcontainerhelper/assets/10775043/1713569d-a3c1-4ff6-bdad-2abe470f2b57)

I tried with combinations of PS5 and PS7 as well - no problems.
I also never saw this on any of my runners, having multiple agents.

**Comment by freddydk** [comment]
In your -asparallel pipeline - it never displays:
`Waiting for other process downloading artifact 'https://bcartifacts.azureedge.net/sandbox/22.4.59114.59980/be`

causing me to believe that the artifacts were cached and never really downloaded.
When I run that command - I get the waiting... line and it is waiting.

**Comment by b-zijlstra** [comment]
- I am running a Windows Server 2022 VM on onprem hardware.
- I cannot reproduce the issue outside of the failing pipeline. I've tried PS5, PS7, with and without running as admin. None have the same issue as within the pipeline.
- In the original issue I had already downloaded the artifacts, so yes, they were cached. I can confirm that the pipeline issue occurs both for the situation where the artifacts are cached, as well as when the artifacts need to be downloaded. If I remove the cache and use powershell twice, then I correctly get the 'Waiting for other process...' message. If I remove the cache and run the pipeline a second time while a previous pipeline is still downloading, then I get the error.
